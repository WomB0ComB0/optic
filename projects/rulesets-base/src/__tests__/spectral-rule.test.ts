import { test, expect, describe } from '@jest/globals';
import { RuleRunner } from '../rule-runner';
import { SpectralRule } from '../extended-rules/spectral-rule';
import {
  Spectral,
  RulesetDefinition as SpectralRulesetDefinition,
} from '@stoplight/spectral-core';
import { oas } from '@stoplight/spectral-rulesets';
import { OpenAPIV3, defaultEmptySpec, diff } from '@useoptic/openapi-utilities';

const spectral = new Spectral();
spectral.setRuleset({
  extends: [[oas as SpectralRulesetDefinition, 'all']],
  rules: {},
});

describe('spectral rule test', () => {
  const ruleRunner = new RuleRunner([
    new SpectralRule({
      name: 'spectral-rules',
      spectral,
    }),
  ]);

  test('runs spectral rules on added', async () => {
    const before: OpenAPIV3.Document = { ...defaultEmptySpec };
    const after: OpenAPIV3.Document = {
      ...defaultEmptySpec,
      paths: {
        '/api': {
          get: {
            responses: {},
          },
        },
      },
    };
    const diffs = diff(before, after);
    const resultsWithChanges = await ruleRunner.runRules({
      diffs,
      fromSpec: before,
      toSpec: after,
      context: {},
    });

    expect(resultsWithChanges.length > 0).toBe(true);
    expect(resultsWithChanges.every((r) => r.passed)).toBe(false);
    expect(resultsWithChanges).toMatchSnapshot();

    const resultsAgainstSelf = await ruleRunner.runRules({
      diffs,
      fromSpec: after,
      toSpec: after,
      context: {},
    });
    expect(resultsAgainstSelf.length === 0).toBe(true);
  });
});
