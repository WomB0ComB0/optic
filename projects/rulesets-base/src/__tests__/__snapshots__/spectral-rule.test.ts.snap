// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spectral rule test runs spectral rules on added 1`] = `
[
  {
    "docsLink": undefined,
    "error": "Error code: oas3-api-servers: OpenAPI "servers" must be present and non-empty array.",
    "exempted": false,
    "location": {
      "jsonPath": "",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: openapi-tags: OpenAPI object must have non-empty "tags" array.",
    "exempted": false,
    "location": {
      "jsonPath": "",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: operation-description: Operation "description" must be present and non-empty string.",
    "exempted": false,
    "location": {
      "jsonPath": "/paths/~1api/get",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: operation-operationId: Operation must have "operationId".",
    "exempted": false,
    "location": {
      "jsonPath": "/paths/~1api/get",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: operation-tags: Operation must have non-empty "tags" array.",
    "exempted": false,
    "location": {
      "jsonPath": "/paths/~1api/get",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: operation-success-response: Operation must have at least one "2xx" or "3xx" response.",
    "exempted": false,
    "location": {
      "jsonPath": "/paths/~1api/get",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: info-contact: Info object must have "contact" object.",
    "exempted": false,
    "location": {
      "jsonPath": "",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: info-description: Info "description" must be present and non-empty string.",
    "exempted": false,
    "location": {
      "jsonPath": "",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: info-license: Info object must have "license" object.",
    "exempted": false,
    "location": {
      "jsonPath": "",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
  {
    "docsLink": undefined,
    "error": "Error code: license-url: License object must include "url".",
    "exempted": false,
    "location": {
      "jsonPath": "",
      "spec": "before",
    },
    "name": "Spectral always rule",
    "passed": false,
    "severity": "warn",
    "type": "requirement",
    "where": "always ",
  },
]
`;
